src/reflection.ts

```diff
@@ -802,18 +811,24 @@ export class Parser {
           continue;
         }
 
-        const deserializer = unionDeserializers.get(discriminator);
-        if (!deserializer) {
+        const specificTypeInfo = specificTypes.get(discriminator);
+        if (!specificTypeInfo) {
           throw new Error(
             `Malformed message: unknown union type '${discriminator}' in field ${discriminatorField.name()}`,
           );
         }
 
-        const subTable = tables[idx];
-        if (!subTable) {
-          throw new Error(`Malformed message missing table at ${field.name()} positon ${idx}`);
-        }
-        result.push(deserializer(subTable));
+        const subTable = new Table(
+          table.bb,
+          specificTypeInfo.typeIndex,
+          table.bb.__indirect(nextTableOffset),
+          false /* elementIsStruct */,
+        );
+
+        // In a vector of union, each "element" is an object until we re-visit struct
+        // support
+        nextTableOffset += typeSize(reflection.BaseType.Obj);
```

Comment:
There might be some general form but I am not sure it would make it _much_ clearer because the current logic is doing the table iteration while going through the discriminator values (rather than reading all of the tables up front) so it kinda feels natural and clear that you are iterating the offset (note you also have to use the correct "specific type index" when building the subtable so it isn't only a matter of iterating with offsets.

Reply:
Fair enough.